Source/WebCore/Modules/permissions/Permissions.cpp

```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+    JSC::VM& vm = context.globalObject()->vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
-    auto parameterDescriptor = convert<IDLDictionary<PermissionDescriptor>>(*context->globalObject(), permissionDescriptorValue.get());
+    auto parameterDescriptor = convert<IDLDictionary<PermissionDescriptor>>(*context.globalObject(), permissionDescriptorValue.get());
     if (UNLIKELY(scope.exception())) {
         promise.reject(Exception { ExistingExceptionError });
         return;
     }
 
-    if (is<Document>(context) && !isAllowedByFeaturePolicy(downcast<Document>(*context), parameterDescriptor.name)) {
-        context->postTask([parameterDescriptor, promise = WTFMove(promise)](auto& context) mutable {
-            promise.resolve(PermissionStatus::create(context, PermissionState::Denied, parameterDescriptor));
-        });
-        return;
+    if (is<Document>(context)) {
+        if (!isAllowedByFeaturePolicy(downcast<Document>(context), parameterDescriptor.name)) {
+            context.postTask([parameterDescriptor, promise = WTFMove(promise)](auto& context) mutable {
+                promise.resolve(PermissionStatus::create(context, PermissionState::Denied, parameterDescriptor));
+            });
+            return;
+        }
     }
 
-    auto* origin = context->securityOrigin();
+    auto* origin = context.securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
 
-    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *document->page(), [this, protectedThis = Ref { *this }, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
-        auto context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
-        if (!context || !context->globalObject())
+    if (is<WorkerGlobalScope>(context)) {
+        auto* workerGlobalScope = downcast<WorkerGlobalScope>(&context);
+
+        auto completionHandler = [originData, parameterDescriptor, identifier = workerGlobalScope->identifier(), promise = WTFMove(promise)] (auto& context) mutable { 
+            PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *downcast<Document>(context).page(), [&context, identifier, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+                if (!context.globalObject())
+                    return;
+
+                context.postTask([identifier, parameterDescriptor, promise = WTFMove(promise), permissionState = WTFMove(permissionState)](auto&) mutable {
+                    ScriptExecutionContext::postTaskTo(identifier, [parameterDescriptor, promise = WTFMove(promise), permissionState = WTFMove(permissionState)](auto& context) mutable {
+                        if (!permissionState)
+                            promise.reject(Exception { NotSupportedError });
```

Comment:
It would be nice to provide a helpful error message as second parameter to the Exception constructor.

Reply:
Fixed!